server/src/storage/datastore.ts

```diff
@@ -22,6 +22,40 @@ import {UpdateRule} from './interfaces';
 
 const datastore = new Datastore();
 
+type TransactionResponse = number | void;
+type TransactionOperation = (
+  transaction: Transaction
+) => Promise<TransactionResponse>;
+
+/**
+ * A higher order function that creates a transaction and carries out all
+ * input datastore operations in the same transaction.
+ * Returns the response of each operation carried out in the transaction in
+ * an array, in the same order they are called.
+ * @param opFns Operations to be carried out in the same transaction
+ */
+export const makeTransaction = async (
+  ...opFns: TransactionOperation[]
+): Promise<TransactionResponse[]> => {
+  const transaction = datastore.transaction();
+
+  const res = [];
+
+  try {
+    await transaction.run();
+
+    for (const fn of opFns) {
```

Comment:
consider renaming fn and opFns, they're not very straightforward.
e.g. operation and operations